### PR TITLE
[Viewer] Make search for X11 window IDs more robust

### PR DESCRIPTION
**Pull request by ejmastnak**
This PR addresses and solves #2415, wherein the variable `b:vimtex.viewer.xwin_id` would sometimes be set to an X11 window ID that was not the ID of the actual viewer. The issue is solved on my end (Arch Linux 5.18.2 with Zathura), but it would be prudent to test on other systems.

The PR is a refactor of existing code more than anything else; here is a quick summary:

- Move winID-searching logic (by viewer PID, name and class) to three dedicated functions (to avoid code repetition; see [5671b1a](https://github.com/lervag/vimtex/commit/5671b1aed1628ea9ff47f966eb3b573b0705ca01) for details)
- Makes the `xdo_get_id` function in `_template.vim` attempt search by PID and search by window name before attempting search by class (previously only search by class was used).

More details in the commit messages.

refer: #2415

### CODE REVIEW

File: autoload/vimtex/view/_template.vim [hunk 1]
```diff
@@ -136,23 +136,37 @@ function! s:viewer.xdo_get_id() dict abort " {{{1
     " Allow some time for the viewer to start properly
```
Comment (lervag) [inline autoload/vimtex/view/_template.vim:136]: Let's simplify further. Something like this:

```vim
if self.xwin_id > 0 | return self.xwin_id | endif

" Allow some time for the viewer to start properly
sleep 500m

" EXPLANATORY COMMENT HERE
for l:method in ['name', 'pid', 'class']
  let self.xwin_id = self.xdo_find_win_id_by_{l:method}()
  if self.xwin_id > 0 | break | endif
endfor

if self.xwin_id <= 0
  call vimtex#log#warning('Viewer cannot find ' . self.name . ' window ID!')
endif

return self.xwin_id
```

File: autoload/vimtex/view/_template.vim [hunk 2]
```diff
@@ -162,31 +176,67 @@ function! s:viewer.xdo_exists() dict abort " {{{1
...
+    if self.xwin_id > 0
+      let self.xwin_id = self.xdo_find_win_id_by_name()
     endif
   endif
 
   return self.xwin_id > 0
 endfunction
 
+" }}}1
+" Attempts to find the viewer's X window ID using `xdotool` search by window
```
Comment (lervag) [inline autoload/vimtex/view/_template.vim:193]: I prefer to have the comments at the top of the function, but not above.

File: autoload/vimtex/view/_template.vim [hunk 3]
```diff
@@ -162,31 +176,67 @@ function! s:viewer.xdo_exists() dict abort " {{{1
...
+" class. Returns the viewer's window ID if one is found, and `0` otherwise.
+function! s:viewer.xdo_find_win_id_by_class() dict abort " {{{1
+  let l:xwin_ids = vimtex#jobs#capture('xdotool search --class ' . self.name)
+  if len(l:xwin_ids) == 0
```
Comment (lervag) [inline autoload/vimtex/view/_template.vim:197]: I would do it like this:

```
return empty(l:xwin_ids) ? 0 : l:xwin_ids[-1]
```

File: autoload/vimtex/view/_template.vim [hunk 4]
```diff
@@ -162,31 +176,67 @@ function! s:viewer.xdo_exists() dict abort " {{{1
...
+  return 0
+endfunction
+
+" }}}1
+" Attempts to find the viewer's X window ID using `xdotool` search by the
+" viewer's process ID. Returns the viewer's window ID if one is found, and `0`
+" otherwise.
+function! s:viewer.xdo_find_win_id_by_pid() dict abort " {{{1
+  let l:pid = has_key(self, 'get_pid') ? self.get_pid() : 0
+  if l:pid > 0
```
Comment (lervag) [inline autoload/vimtex/view/_template.vim:230]: I would simplify like this:

```vim
if l:pid <= 0 | return 0 | endif

let xwin_ids = vimtex#jobs#capture(
      \   'xdotool search --all --pid ' . l:pid
      \ . ' --name ' . fnamemodify(self.out(), ':t'))
return get(xwin_ids, 0)
```

**Comment by lervag** [comment]
Thanks! I think this looks very good and is almost ready for merge already. I only had a few minor comments and suggestions.

File: autoload/vimtex/view/_template.vim (hunk 1 above)
Comment (ejmastnak) [reply autoload/vimtex/view/_template.vim:136]: Very elegant! Note that the code...
```vim
let self.xwin_id = self.xdo_find_win_id_by_{l:method}()
```
...give me the error `E716: Key not present in Dictionary: "xdo_find_win_id_by_{l:method}()"`.

I've taken the liberty of replacing it with
```vim
execute "let self.xwin_id = self.xdo_find_win_id_by_" . l:method . "()"
```
which I believe does the trick.

File: autoload/vimtex/view/_template.vim (hunk 3 above)
Comment (ejmastnak) [reply autoload/vimtex/view/_template.vim:197]: Nice, I like the C-style conditional `?:`.
Will update.

**Comment by lervag** [comment]
Thanks! I made a few adjustments on top, I hope you don't mind. Let me know if I accidentily broke something :p

**Comment by ejmastnak** [comment]
> Thanks! I made a few adjustments on top, I hope you don't mind. Let me know if I accidentily broke something :p

Looks good! Just tested locally and everything seems to work. Thanks for your suggestions and help implementing this!

**Pull request merged**